File: day09.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INPUT "day09.input"

void part1n2(char *line)
{
    char c;
    int i, garbage, ignore, group, score, garbage_score;
    garbage = ignore = group = score = garbage_score = 0;

    for (i = 0; i < strlen(line); i++) {
        c = line[i];
        if (garbage) {
            if (ignore)
                ignore = 0;
            else if (c == '!')
                ignore = 1;
            else if (c == '>')
                garbage = 0;
            else
                garbage_score += 1;
        } else {
            if (c == '<')
                garbage = 1;
            else if (c == '{')
                group += 1;
            else if (c == '}') {
                score += group;
                group -= 1;
            }
        }
    }

    printf("%d %d\n", score, garbage_score);
}

int main()
{
    FILE *fp;
    char *line = NULL;
    size_t len = 0;

    fp = fopen(INPUT, "r");
    if (fp == NULL) {
        perror(INPUT);
        exit(1);
    }

    while (getline(&line, &len, fp) != -1) {
        line[strcspn(line, "\r\n")] = 0;
        part1n2(line);
    }

    return 0;
}

File: day09.py
#!/usr/bin/env python

with open('day09.input') as f:
    data = f.read().rstrip()

def part1n2():
    score = group = junk = 0
    garbage = ignore = False

    for i in range(len(data)):
        c = data[i]

        if garbage:
            if ignore:
                ignore = False
            elif c == '!':
                ignore = True
            elif c == '>':
                garbage = False
            else:
                junk += 1
        else:
            if c == '<':
                garbage = True
            elif c == '{':
                group += 1
            elif c == '}':
                score += group
                group -= 1

    return score, junk

print(part1n2())
